File: epi_judge_python/power_x_y.py
from test_framework import generic_test


def power(x, y):
    result, pow = 1, y
    if y < 0:
        pow, x = -pow, 1.0 / x

    while pow:
        if pow & 1:
            pow -= 1
            result *= x
        else:
            pow >>= 1
            x *= x
    return result

if __name__ == '__main__':
    exit(
        generic_test.generic_test_main("power_x_y.py", 'power_x_y.tsv', power))

File: epi_judge_java/epi/PowerXY.java
package epi;
import epi.test_framework.EpiTest;
import epi.test_framework.GenericTest;
public class PowerXY {
  @EpiTest(testDataFile = "power_x_y.tsv")
  public static double power(double x, int y) {
    if (y == 1){
      return x;
    } else if (y == 0){
      return 1;

    } else if (y < 0) {
      return 1 / power(x,-y);
    }

    if (y % 2 == 0){
      return power(x*x, y/2);
    } else {
      return x * power(x, y - 1);
    }
  }

  public static void main(String[] args) {
    System.exit(
        GenericTest
            .runFromAnnotations(args, "PowerXY.java",
                                new Object() {}.getClass().getEnclosingClass())
            .ordinal());
  }
}
